include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +171,124 @@ class fast_istreambuf_iterator
+     */
+    fast_ostreambuf_iterator() noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator(fast_ostreambuf_iterator const &) noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator(fast_ostreambuf_iterator &&) noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator & operator=(fast_ostreambuf_iterator const &) noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator & operator=(fast_ostreambuf_iterator &&) noexcept = default; //!< Defaulted.
+    ~fast_ostreambuf_iterator() noexcept = default; //!< Defaulted.
+
+    //!\brief Construct from a stream buffer.
+    explicit fast_ostreambuf_iterator(std::basic_streambuf<char_t, traits_t> & ibuf) :
+        stream_buf{reinterpret_cast<stream_buffer_exposer<char_t, traits_t> *>(&ibuf)}
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+            stream_buf->overflow(); // ensures that put area has space available
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator++()
+    {
+        return *this;
+    }
+    //!\overload
+    fast_ostreambuf_iterator & operator++(int)
+    {
+        return *this;
+    }
+    //!\}
+
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator*()
+    {
+        return *this;
+    }
+
+    //!\brief Writes a character to the associated output stream.
+    fast_ostreambuf_iterator & operator=(char_t const c)
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+        {
+            if (stream_buf->sputc(c) == traits_t::eof()) // overflow() [virtual], then write character
+                throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+        }
+        else
+        {
+            *stream_buf->pptr() = c;
+            stream_buf->pbump(1); // advance pptr() in put area without any checks
+        }
+        return *this;
+    }
+
+    //!\brief Returns `true if this iterator has encountered the end-of-file condition on output, `false` otherwise.
+    bool failed() const noexcept
+    {
+        return stream_buf->overflow() == traits_t::eof();
+    }
+
+    /*!\brief Writes a range to the associated output.
+     * \tparam range_type The type of range to write; Must model std::ranges::forward_range.
+     * \param[in] rng The range to write.
+     */
+    template <std::ranges::forward_range range_type>
+    auto write_range(range_type && rng)
+    {
+        auto buffer_space = stream_buf->epptr() - stream_buf->pptr();
+        auto copy_res = std::ranges::copy(rng | views::take(buffer_space), stream_buf->pptr());
+        auto written_chars = copy_res.out - stream_buf->pptr();
+        stream_buf->pbump(written_chars);
+
+        auto remaining_range = rng | views::drop(written_chars);
+
+        if (remaining_range.begin() == remaining_range.end()) // no more characters to write
```

Comment:
```suggestion
        if (copy_res.in == std::ranges::end(rng)) // no more characters to write
```

Reply:
This might not work because the input to `std::ranges::copy` is `rng | take(..)` so the iretors are not comparable :(

Reply:
Ah, correct! But You can do :
```cpp
auto tview = rng | views::take(buffer_space);
auto copy_res = std::ranges::copy(tview, stream_buf->pptr());
// ...
if (copy_res.in == std::ranges::end(tview)) // no more characters to write
```

Or, not?

Reply:
Oh, you are right of course! However, this means that you are also not returning an iterator of the original range when calling `return copy_res.in;` and outside this function this will be dangling (it refers to the take_view) and you are assigning it back to the original iterator :boom: (This all just works because our iterators inherit from each other and casting actually does the correct thing by chance)

I would recommend just doing what I proposed on the outside, as well: 

```cpp
                 auto it = std::ranges::begin(rng);
                 auto e = it;
                 std::ranges::advance(e, buffer_space, std::ranges::end(v));
                 auto copy_res = std::ranges::copy(it, e, stream_buf->pptr());
              
                 // ...

                 if (copy_res.in == std::ranges::end(rng)) // no more characters to write
```

Now `copy_res.in` is guaranteed to be of the original iterator type. These changes should have zero overhead for sized random access ranges (advance is in O(1) for them).

Reply:
However, this functions should also be documented that if an rvalue of a range is passed to it and that range is not a safe_range than any returned iterator will still be dangling.